File: web/src/lib/tool-workflow-observability.ts
import type { ToolExposureDiagnostics } from './chat.ts'

export interface EffectivePolicyProjection {
  policySnapshotId: string | null
  policyVersion: string | null
  sourceChain: string[]
  autonomyMode: string
  requireApprovalForFileWrite: boolean | null
  requireApprovalForExec: boolean | null
  fileReadScope: string | null
  fileWriteScope: string | null
  hardDenies: string[]
  expectationStatus: 'matches' | 'stale' | 'not_provided'
  expectedPolicyVersion: string | null
  reviewSemantics: 'concrete_call_only'
}

export interface ToolWorkflowCallProjection {
  callId: string
  toolName: string
  status: string
  arguments?: Record<string, unknown>
  target?: string | null
  approvalId?: string | null
  approvalStatus?: string
  autoReviewDecision?: string
  autoReviewSource?: string | null
  operationId?: string | null
  changedPaths?: string[]
  verificationStatus?: string
  retryStatus?: string
  blocker?: string | null
}

export interface ToolWorkflowProjection {
  schemaVersion: number
  effectivePolicy: EffectivePolicyProjection
  toolInventory: {
    catalogScope: 'policy_eligible'
    policyCatalog: string[]
    eligibleTools: string[]
    exposedTools: string[]
  }
  activation: { status: string; calls: ToolWorkflowCallProjection[] }
  callReview: { status: string; calls: ToolWorkflowCallProjection[] }
  execution: { status: string; calls: ToolWorkflowCallProjection[] }
}

export interface ConcreteToolWorkflowView {
  activationStatus: string | null
  activationAuthorizesCall: boolean | null
  reviewStatus: string
  approvalStatus: string
  autoReviewDecision: string
  autoReviewSource: string | null
  executionStatus: string
  operationId: string | null
  changedPaths: string[]
  verificationStatus: string
  retryStatus: string
  blocker: string | null
}

function record(value: unknown): Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    ? value as Record<string, unknown>
    : {}
}

function records(value: unknown): Record<string, unknown>[] {
  return Array.isArray(value) ? value.map(record).filter((item) => Object.keys(item).length > 0) : []
}

function text(value: unknown): string | null {
  return typeof value === 'string' && value.trim().length > 0 ? value.trim() : null
}

function strings(value: unknown): string[] {
  return Array.isArray(value)
    ? value.filter((item): item is string => typeof item === 'string' && item.length > 0)
    : []
}

function optionalBoolean(...values: unknown[]): boolean | null {
  for (const value of values) {
    if (typeof value === 'boolean') return value
  }
  return null
}

function capabilityPlan(exposure: Record<string, unknown>): Record<string, unknown> {
  const diagnostics = record(exposure.diagnostics)
  const stages = records(diagnostics.stages)
  for (let index = stages.length - 1; index >= 0; index -= 1) {
    const stage = stages[index]
    if (stage.stage === 'turn_contract_rollout') {
      return record(stage.capability_plan)
    }
  }
  return {}
}

export function normalizeToolExposureDiagnostics(
  metadata?: Record<string, unknown>
): ToolExposureDiagnostics | undefined {
  const exposure = record(metadata?.tool_exposure ?? metadata?.toolExposure)
  if (Object.keys(exposure).length === 0) return undefined

  const plan = capabilityPlan(exposure)
  const rawDiagnostics = records(plan.tool_diagnostics)
  const adapter = record(record(exposure.diagnostics).capability_exposure_adapter)
  const broker = record(adapter.activation_broker)
  const exposedTools = strings(exposure.exposed_tools ?? exposure.exposedTools)
  const eligibleTools = strings(plan.eligible_tools)
  const policyCatalog = rawDiagnostics
    .map((item) => text(item.tool_name))
    .filter((item): item is string => item !== null)
  const workspaceBound = optionalBoolean(exposure.workspace_bound, exposure.workspaceBound)
  const attachmentCount = [exposure.attachment_count, exposure.attachmentCount]
    .find((value): value is number => typeof value === 'number' && Number.isFinite(value))

  return {
    catalogScope: 'policy_eligible',
    policyCatalog,
    eligibleTools,
    exposedTools,
    activationAllowedTools: strings(adapter.activation_allowed_tool_names),
    deferredTools: strings(broker.deferred_tool_names),
    planVersion: text(plan.plan_version) ?? text(adapter.plan_version) ?? undefined,
    toolDiagnostics: rawDiagnostics.map((item) => ({
      toolName: text(item.tool_name) ?? 'unknown',
      status: text(item.status) ?? 'unknown',
      includeReasons: strings(item.include_reasons),
      excludeReasons: strings(item.exclude_reasons),
    })),
    workspaceBound: workspaceBound ?? undefined,
    attachmentCount,
  }
}

function normalizeCall(value: unknown): ToolWorkflowCallProjection | null {
  const item = record(value)
  const callId = text(item.call_id)
  if (!callId) return null
  return {
    callId,
    toolName: text(item.tool_name) ?? text(item.requested_tool) ?? 'tool_activate',
    status: text(item.status) ?? 'not_observed',
    arguments: Object.keys(record(item.arguments)).length > 0 ? record(item.arguments) : undefined,
    target: text(item.target),
    approvalId: text(item.approval_id),
    approvalStatus: text(item.approval_status) ?? undefined,
    autoReviewDecision: text(item.auto_review_decision) ?? undefined,
    autoReviewSource: text(item.auto_review_source),
    operationId: text(item.operation_id),
    changedPaths: strings(item.changed_paths),
    verificationStatus: text(item.verification_status) ?? undefined,
    retryStatus: text(item.retry_status) ?? undefined,
    blocker: text(item.blocker),
  }
}

function normalizeCallGroup(value: unknown): { status: string; calls: ToolWorkflowCallProjection[] } {
  const group = record(value)
  return {
    status: text(group.status) ?? 'not_observed',
    calls: (Array.isArray(group.calls) ? group.calls : [])
      .map(normalizeCall)
      .filter((item): item is ToolWorkflowCallProjection => item !== null),
  }
}

export function normalizeToolWorkflowProjection(value: unknown): ToolWorkflowProjection | null {
  const root = record(value)
  if (Object.keys(root).length === 0) return null
  const policy = record(root.effective_policy)
  const inventory = record(root.tool_inventory)
  const expectation = text(policy.expectation_status)
  return {
    schemaVersion: typeof root.schema_version === 'number' ? root.schema_version : 1,
    effectivePolicy: {
      policySnapshotId: text(policy.policy_snapshot_id),
      policyVersion: text(policy.policy_version),
      sourceChain: strings(policy.source_chain),
      autonomyMode: text(policy.autonomy_mode) ?? 'unknown',
      requireApprovalForFileWrite: optionalBoolean(policy.require_approval_for_file_write),
      requireApprovalForExec: optionalBoolean(policy.require_approval_for_exec),
      fileReadScope: text(policy.file_read_scope),
      fileWriteScope: text(policy.file_write_scope),
      hardDenies: strings(policy.hard_denies),
      expectationStatus: expectation === 'matches' || expectation === 'stale' ? expectation : 'not_provided',
      expectedPolicyVersion: text(policy.expected_policy_version),
      reviewSemantics: 'concrete_call_only',
    },
    toolInventory: {
      catalogScope: 'policy_eligible',
      policyCatalog: strings(inventory.policy_catalog),
      eligibleTools: strings(inventory.eligible_tools),
      exposedTools: strings(inventory.exposed_tools),
    },
    activation: normalizeCallGroup(root.activation),
    callReview: normalizeCallGroup(root.call_review),
    execution: normalizeCallGroup(root.execution),
  }
}

export function projectConcreteToolWorkflow(input: {
  toolName: string
  result?: unknown
  metadata?: Record<string, unknown>
  isResult: boolean
  errorMessage?: string
}): ConcreteToolWorkflowView {
  const metadata = record(input.metadata)
  const result = record(input.result)
  const isActivation = input.toolName === 'tool_activate'
  const activationStatus = isActivation
    ? text(metadata.status) ?? text(result.status) ?? (input.errorMessage ? 'failed' : 'not_observed')
    : null
  const approvalStatus = text(metadata.approval_status) ?? text(metadata.approval_state) ?? 'not_observed'
  const autoReviewDecision = text(metadata.auto_review_decision) ?? 'not_observed'
  return {
    activationStatus,
    activationAuthorizesCall: isActivation
      ? optionalBoolean(
          metadata.activation_authorizes_tool_call,
          result.activation_authorizes_tool_call
        )
      : null,
    reviewStatus: autoReviewDecision !== 'not_observed' ? autoReviewDecision : approvalStatus,
    approvalStatus,
    autoReviewDecision,
    autoReviewSource: text(metadata.auto_review_source),
    executionStatus: input.isResult ? (input.errorMessage ? 'failed' : 'completed') : 'not_observed',
    operationId: text(metadata.operation_id) ?? text(result.operation_id),
    changedPaths: strings(metadata.changed_paths ?? result.changed_paths),
    verificationStatus: text(metadata.verification_status) ?? text(result.verification_status) ?? 'not_observed',
    retryStatus: text(metadata.retry_status) ?? 'not_observed',
    blocker: text(metadata.blocker) ?? text(metadata.reason),
  }
}

File: mochi/api/tool_workflow_observability.py
"""Read-only projection of ordinary Chat tool workflow state."""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any


def build_tool_workflow_observability(
    *,
    events: Sequence[Mapping[str, Any]],
    effective_policy: Mapping[str, Any],
    expected_policy_version: str | None = None,
) -> dict[str, Any]:
    """Project authoritative runtime facts without changing workflow state."""

    policy_version = _text(effective_policy.get("policy_version"))
    policy_expectation = "not_provided"
    if expected_policy_version:
        policy_expectation = (
            "matches" if expected_policy_version == policy_version else "stale"
        )

    exposure = _latest_tool_exposure(events)
    capability_plan = _capability_plan(exposure)
    tool_diagnostics = _record_list(capability_plan.get("tool_diagnostics"))
    policy_catalog = [
        name
        for item in tool_diagnostics
        if (name := _text(item.get("tool_name"))) is not None
    ]

    request_arguments: dict[str, dict[str, Any]] = {}
    activation_by_call: dict[str, dict[str, Any]] = {}
    execution_by_call: dict[str, dict[str, Any]] = {}
    review_by_call: dict[str, dict[str, Any]] = {}

    for index, event in enumerate(events):
        event_type = _text(event.get("type")) or ""
        call_id = _text(event.get("call_id")) or f"event-{index}"
        tool_name = _text(event.get("tool_name")) or ""
        arguments = _record(event.get("arguments"))
        if arguments:
            request_arguments[call_id] = arguments
        if event_type not in {"tool_call_completed", "tool_call_result"}:
            continue

        metadata = _record(event.get("metadata"))
        result = _record(event.get("result"))
        effective_arguments = arguments or request_arguments.get(call_id, {})
        if tool_name == "tool_activate":
            activation_by_call[call_id] = _activation_projection(
                call_id=call_id,
                metadata=metadata,
                result=result,
                arguments=effective_arguments,
                error=_text(event.get("error")),
            )
            continue

        review_by_call[call_id] = _review_projection(
            call_id=call_id,
            tool_name=tool_name,
            arguments=effective_arguments,
            metadata=metadata,
        )
        execution_by_call[call_id] = _execution_projection(
            call_id=call_id,
            tool_name=tool_name,
            metadata=metadata,
            result=result,
            error=_text(event.get("error")),
        )

    exposed_tools = _string_list(exposure.get("exposed_tools"))
    eligible_tools = _string_list(capability_plan.get("eligible_tools"))
    return {
        "schema_version": 1,
        "effective_policy": {
            **dict(effective_policy),
            "expectation_status": policy_expectation,
            "expected_policy_version": expected_policy_version,
            "review_semantics": "concrete_call_only",
        },
        "tool_inventory": {
            "catalog_scope": "policy_eligible",
            "policy_catalog": policy_catalog,
            "eligible_tools": eligible_tools,
            "exposed_tools": exposed_tools,
            "tool_diagnostics": tool_diagnostics,
        },
        "activation": {
            "status": "observed" if activation_by_call else "not_observed",
            "calls": list(activation_by_call.values()),
        },
        "call_review": {
            "status": "observed" if review_by_call else "not_observed",
            "calls": list(review_by_call.values()),
        },
        "execution": {
            "status": "observed" if execution_by_call else "not_observed",
            "calls": list(execution_by_call.values()),
        },
    }


def _latest_tool_exposure(events: Sequence[Mapping[str, Any]]) -> dict[str, Any]:
    for event in reversed(events):
        metadata = _record(event.get("metadata"))
        exposure = _record(metadata.get("tool_exposure"))
        if exposure:
            return exposure
    return {}


def _capability_plan(exposure: Mapping[str, Any]) -> dict[str, Any]:
    diagnostics = _record(exposure.get("diagnostics"))
    for stage in reversed(_record_list(diagnostics.get("stages"))):
        if stage.get("stage") != "turn_contract_rollout":
            continue
        plan = _record(stage.get("capability_plan"))
        if plan:
            return plan
    return {}


def _activation_projection(
    *,
    call_id: str,
    metadata: Mapping[str, Any],
    result: Mapping[str, Any],
    arguments: Mapping[str, Any],
    error: str | None,
) -> dict[str, Any]:
    status = _text(metadata.get("status")) or _text(result.get("status"))
    return {
        "call_id": call_id,
        "requested_tool": (
            _text(metadata.get("requested_tool"))
            or _text(result.get("requested_tool"))
            or _text(arguments.get("tool_name"))
        ),
        "status": status or ("failed" if error else "not_observed"),
        "reason": _text(metadata.get("reason")) or _text(metadata.get("error_type")),
        "callable_this_turn": _optional_bool(
            metadata.get("callable_this_turn"),
            result.get("callable_this_turn"),
        ),
        "activation_authorizes_tool_call": _optional_bool(
            metadata.get("activation_authorizes_tool_call"),
            result.get("activation_authorizes_tool_call"),
        ),
    }


def _review_projection(
    *,
    call_id: str,
    tool_name: str,
    arguments: Mapping[str, Any],
    metadata: Mapping[str, Any],
) -> dict[str, Any]:
    auto_review_decision = _text(metadata.get("auto_review_decision"))
    approval_status = _text(metadata.get("approval_status")) or _text(
        metadata.get("approval_state")
    )
    review_status = auto_review_decision or approval_status or "not_observed"
    return {
        "call_id": call_id,
        "tool_name": tool_name,
        "arguments": dict(arguments),
        "target": _target_projection(arguments, metadata),
        "status": review_status,
        "approval_id": _text(metadata.get("approval_id")),
        "approval_status": approval_status or "not_observed",
        "auto_review_decision": auto_review_decision or "not_observed",
        "auto_review_source": _text(metadata.get("auto_review_source")),
    }


def _execution_projection(
    *,
    call_id: str,
    tool_name: str,
    metadata: Mapping[str, Any],
    result: Mapping[str, Any],
    error: str | None,
) -> dict[str, Any]:
    verification_status = (
        _text(metadata.get("verification_status"))
        or _text(result.get("verification_status"))
        or "not_observed"
    )
    return {
        "call_id": call_id,
        "tool_name": tool_name,
        "status": "failed" if error else "completed",
        "operation_id": _text(metadata.get("operation_id"))
        or _text(result.get("operation_id")),
        "changed_paths": _string_list(
            metadata.get("changed_paths") or result.get("changed_paths")
        ),
        "verification_status": verification_status,
        "retry_status": _text(metadata.get("retry_status")) or "not_observed",
        "blocker": _text(metadata.get("blocker")) or _text(metadata.get("reason")),
    }


def _target_projection(
    arguments: Mapping[str, Any], metadata: Mapping[str, Any]
) -> str | None:
    for key in ("resolved_target", "target", "path", "file_path", "workdir", "url"):
        value = _text(metadata.get(key)) or _text(arguments.get(key))
        if value:
            return value
    return None


def _record(value: Any) -> dict[str, Any]:
    return dict(value) if isinstance(value, Mapping) else {}


def _record_list(value: Any) -> list[dict[str, Any]]:
    if not isinstance(value, list):
        return []
    return [_record(item) for item in value if isinstance(item, Mapping)]


def _string_list(value: Any) -> list[str]:
    if not isinstance(value, (list, tuple)):
        return []
    return [item for item in value if isinstance(item, str) and item]


def _text(value: Any) -> str | None:
    return value.strip() if isinstance(value, str) and value.strip() else None


def _optional_bool(*values: Any) -> bool | None:
    for value in values:
        if isinstance(value, bool):
            return value
    return None
